src/go/nerdctl-stub/command_handlers.go

```diff
@@ -22,15 +25,133 @@ func imageBuildHandler(c *commandDefinition, args []string) (*parsedArgs, error)
+			err = multierror.Append(err, cleanupErr)
 		}
 		return nil, err
 	}
 	return &parsedArgs{args: append([]string{newPath}, args[1:]...), cleanup: cleanups}, nil
 }
+
+// hostPathResult is the return value of a hostPathDeterminerFunc that is used
+// in containerCopyHandler for determining which argument is the host path that
+// must be munged.
+type hostPathResult int
+
+const (
+	hostPathUnknown = hostPathResult(iota)
+	hostPathCurrent = hostPathResult(iota)
+	hostPathOther   = hostPathResult(iota)
+	hostPathNeither = hostPathResult(iota)
+)
+
+// containerCopyHandler handles `nerdctl container cp`
+func containerCopyHandler(c *commandDefinition, args []string, argHandlers argHandlersType) (*parsedArgs, error) {
+	var resultArgs []string
+	var cleanups []cleanupFunc
+	var paths []string
+
+	// Positional arguments `nerdctl container cp` are all paths, whether inside
+	// the container or outside.
+
+	for _, arg := range args {
+		if arg == "-" || !strings.HasPrefix(arg, "-") {
+			// If the arg is "-" (stdin/stdout) or doesn't start with -, it's a path.
+			paths = append(paths, arg)
+		} else {
+			resultArgs = append(resultArgs, arg)
+		}
+	}
+
+	if len(paths) != 2 {
+		// We should have exactly one source and one destination... just fail
+		err := fmt.Errorf("accepts 2 args, received %d", len(paths))
+		if cleanupErr := runCleanups(cleanups); cleanupErr != nil {
+			err = multierror.Append(err, cleanupErr)
+		}
+		return nil, err
+	}
+
+	hostPathDeterminerFuncs := []func(i int, p string) hostPathResult{
+		func(i int, p string) hostPathResult {
+			if p == "-" {
+				// If one argument is "-", the other must be a container path, so
+				// neither needs to be modified.
+				return hostPathNeither
+			}
+			return hostPathUnknown
+		},
+		func(i int, p string) hostPathResult {
+			colon := strings.Index(p, ":")
+			if colon < 1 {
+				// If there's no colon in the path specification at all, or if the
+				// string starts with a colon (which is invalid), then this must not be
+				// a container path (and therefore the other one is).
+				return hostPathCurrent
+			}
+			return hostPathUnknown
+		},
+		func(i int, p string) hostPathResult {
+			colon := strings.Index(p, ":")
+			if colon > 1 {
+				// There's multiple characters before the first colon; this is a container
+				// path specification (foo:/path/in/container), so the other must be a
+				// host path specification.
+				return hostPathOther
+			}
+			return hostPathUnknown
+		},
+		func(i int, p string) hostPathResult {
+			colon := strings.Index(p, ":")
+			if colon != 1 {
+				// Shouldn't get here -- one of the two previous functions should have
+				// found something already.
+				panic(fmt.Sprintf("Expected path %q to start with a character followed by a colon!", p))
+			}
+			// Fall back: the first element should be treated as the container path.
+			if i == 0 {
+				return hostPathOther
+			}
+			return hostPathCurrent
+		},
```

Comment:
I feel that this whole mechanism of a list of functions is more complex than it needs to be. Writing it as a single function seems much easier to understand:

```golang
		func(i int, p string) hostPathResult {
			if p == "-" {
				// If one argument is "-", the other must be a container path, so
				// neither needs to be modified.
				return hostPathNeither
			}
			colon := strings.Index(p, ":")
			switch {
			case colon < 1:
				// If there's no colon in the path specification at all, or if the
				// string starts with a colon (which is invalid), then this must not be
				// a container path (and therefore the other one is).
				return hostPathCurrent
			case colon > 1:
				// There's multiple characters before the first colon; this is a container
				// path specification (foo:/path/in/container), so the other must be a
				// host path specification.
				return hostPathOther
			default: // colon == 1
				// Fall back: the first element should be treated as the container path.
				if i == 0 {
					return hostPathOther
				}
				return hostPathCurrent
			}
		},
```

And you can get rid of `hostPathUnknown`, which is no longer needed.

We can leave it as-is to get the code merged, but otherwise I prefer the simplified form (not as a single entry function list, but as a standalone function, of course).

Reply:
Sorry, I think I'm wrong; please disregard. The code is not equivalent.